dali/image/tiff.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/image/tiff.h"
+
+namespace dali {
+
+namespace {
+
+const std::vector<int> header_intel = {77, 77, 0, 42};
+const std::vector<int> header_motorola = {73, 73, 42, 0};
+const int COUNT_SIZE = 2;
+const int ENTRY_SIZE = 12;
+const int WIDTH_TAG = 256;
+const int HEIGHT_TAG = 257;
+const int TYPE_WORD = 3;
+const int TYPE_DWORD = 4;
+
+
+cv::Mat DecodeTiff(const unsigned char *tiff, int size, DALIImageType image_type) {
+    std::vector<char> buf(tiff, tiff + size);
+    return cv::imdecode(buf, image_type == DALI_GRAY ? 0 : 1);
+}
+
+
+bool check_header(const unsigned char *tiff, const std::vector<int> &header) {
+    for (unsigned int i = 0; i < header.size(); i++) {
+        if (tiff[i] != header[i]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+
+bool is_little_endian(const unsigned char *tiff) {
+    return check_header(tiff, header_intel);
+}
+
+}  // namespace
+
+bool CheckIsTiff(const unsigned char *tiff) {
+    assert(tiff);
```

Comment:
do we prefer DALI_ASSERT or regular assert?

Reply:
I think DALI_ASSERT is preferred, 'assert' seems to appear due to accident in the code.

Reply:
DALI_ENFORCE(tiff != nullptr) ?

Reply:
Wow, didn't expect that...

So turns out, that what `assert(...)` does is something quite different than `DALI_ASSERT`, however is almost the same as `DALI_ENFORCE`.
But I'd still opt for good old `assert`, because it's turned off by `NDEBUG` flag (which is further set by cmake for Release confs). That way, we have some checks during development, for the errors, that are caused by our misundestandings, not because something went wrong in runtime. And we provide no overhead in actual release

Reply:
For me, turning it off (by it being assert and not DALI_ENFORCE) is premature optimization (please provide profiling results and then we can discuss if this is a big issue or not).
Think about this: if this is DALI_ENFORCE and some bug in DALI will make this a `nullptr`, user will get an error message about `tiff` being `nullptr`, comes to our issues and immediately gives us some useful information. If this is `assert` that is not present in release build, the same user gets just a segfault that does not give much information (especially since they don't know if it is DALI that segfaulted or a framework for example).